pkg/sql/plan/agg_pushdown_pullup.go

```diff
@@ -0,0 +1,153 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import "github.com/matrixorigin/matrixone/pkg/pb/plan"
+
+// some restrictions for agg pushdown to make it easier to acheive
+// will remove some restrictions in the future
+func shouldAggPushDown(agg, join, leftChild, rightChild *plan.Node, builder *QueryBuilder) bool {
+	if leftChild.NodeType != plan.Node_TABLE_SCAN || rightChild.NodeType != plan.Node_TABLE_SCAN {
+		return false
+	}
+	if len(agg.GroupBy) != 0 {
+		return false
+	}
+	if len(agg.AggList) != 1 {
+		return false
+	}
+	aggFunc, ok := agg.AggList[0].Expr.(*plan.Expr_F)
+	if !ok {
+		return false
+	}
+	if aggFunc.F.Func.ObjName != "sum" {
+		return false
+	}
+	colAgg, ok := aggFunc.F.Args[0].Expr.(*plan.Expr_Col)
+	if !ok {
+		return false
+	}
+	leftChildTag := leftChild.BindingTags[0]
+	if colAgg.Col.RelPos != leftChildTag {
+		return false
+	}
+
+	if !IsEquiJoin(join.OnList) || len(join.OnList) != 1 {
+		return false
+	}
+	colGroupBy, ok := filterTag(join.OnList[0], leftChildTag).Expr.(*plan.Expr_Col)
+	if !ok {
+		return false
+	}
+	ndv := getColNdv(colGroupBy.Col, join.NodeId, builder)
+	if ndv < 0 || ndv > join.Stats.Outcnt {
+		return false
+	}
+	return true
+}
+
+func replaceCol(expr *plan.Expr, oldRelPos, oldColPos, newRelPos, newColPos int32) {
+	if expr == nil {
+		return
+	}
+	switch exprImpl := expr.Expr.(type) {
+	case *plan.Expr_F:
+		for _, arg := range exprImpl.F.Args {
+			replaceCol(arg, oldRelPos, oldColPos, newRelPos, newColPos)
+		}
+
+	case *plan.Expr_Col:
+		if exprImpl.Col.RelPos == oldRelPos {
+			exprImpl.Col.RelPos = newRelPos
+			exprImpl.Col.ColPos = 0
+		}
+	}
+}
+
+func filterTag(expr *Expr, tag int32) *Expr {
+	switch exprImpl := expr.Expr.(type) {
+	case *plan.Expr_F:
+		for _, arg := range exprImpl.F.Args {
+			retExpr := filterTag(arg, tag)
+			if retExpr != nil {
+				return retExpr
+			}
+		}
+	case *plan.Expr_Col:
+		if exprImpl.Col.RelPos == tag {
+			return expr
+		}
+	}
+	return nil
+}
+
+func createNewAggNode(agg, join, leftChild *plan.Node, builder *QueryBuilder) {
+	leftChildTag := leftChild.BindingTags[0]
+	newAggList := DeepCopyExprList(agg.AggList)
+	//newGroupBy := DeepCopyExprList(agg.GroupBy)
+	newGroupBy := []*plan.Expr{DeepCopyExpr(filterTag(join.OnList[0], leftChildTag))}
+
+	newGroupTag := builder.genNewTag()
+	newAggTag := builder.genNewTag()
+	newNodeID := builder.appendNode(
+		&plan.Node{
+			NodeType:    plan.Node_AGG,
+			Children:    []int32{leftChild.NodeId},
+			GroupBy:     newGroupBy,
+			AggList:     newAggList,
+			BindingTags: []int32{newGroupTag, newAggTag},
+		},
+		builder.ctxByNode[join.NodeId])
+
+	newBinding := builder.ctxByNode[newNodeID].bindingByTag
+	newBinding[newGroupTag] = newBinding[leftChildTag]
+	newBinding[newAggTag] = newBinding[leftChildTag]
```

Comment:
These lines are wrong. Their is no "binding" attached to the group-by list and aggregate list.